### ISSUE TITLE
[NEXT-901] Jest configuration generated with `next/jest` is not JSON-serializable

### ISSUE THREAD

**Issue by cprussin** [vercel/next.js#47407]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
❯ ./node_modules/.bin/next info

    Operating System:
      Platform: linux
      Arch: x64
      Version: #1-NixOS SMP Mon Mar 13 09:20:37 UTC 2023
    Binaries:
      Node: 18.14.1
      npm: 9.3.1
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.2.5-canary.12
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Jest (next/jest)

### Link to the code that reproduces this issue

https://github.com/cprussin/create-jest-runner-issue-repro

### To Reproduce

- Clone repo
- `npm install`
- `npm exec jest`

### Describe the Bug

While investigating https://github.com/jest-community/create-jest-runner/issues/201, I discovered that `next/jest` generates a jest config that is not JSON-serializable [by attaching the next config to the transformer config](https://github.com/vercel/next.js/blob/bbd79ac9977e03010e09ff90d933f2ec614b031f/packages/next/src/build/jest/jest.ts#L144).

The Jest docs [clearly state](https://jestjs.io/docs/configuration) that the config should always be JSON-serializable, and violating that results in tasks run with [Jest's child process worker](https://github.com/facebook/jest/blob/main/packages/jest-worker/src/workers/ChildProcessWorker.ts) failing with the error message `()=>null could not be cloned.` (basically the serialization to send the job to the child process fails).

The most consistent way I've found to reproduce the issue is to use something based on `create-jest-runner`, such as `jest-runner-eslint`, alongside the next jest runner, but it's most certainly not the only way to get jest to put work on the child process worker (I'm a bit hazy on the details of exactly what causes jest to use the child process worker and what doesn't TBH).

The simplest fix here is to strip non-serializable elements from the next config before passing it to the transformer [here](https://github.com/vercel/next.js/blob/bbd79ac9977e03010e09ff90d933f2ec614b031f/packages/next/src/build/jest/jest.ts#L144).  In my testing, it seems like `generateBuildId` is the function that causes the problem for my minimal repro case.  However, I'm sure there are other functions that will make their way into the config and cause problems, so it may be more appropriate to create an allowlist of parts of the next config that get passed down to the swc jest transformer rather than trying to delete non-serializable elements out of it.

### Expected Behavior

The jest config generated by `next/jest` should be serializable and should not cause `()=>null could not be cloned.` when sending workers to a child process.

<sub>[NEXT-901](https://linear.app/vercel/issue/NEXT-901/jest-configuration-generated-with-nextjest-is-not-json-serializable)</sub>

**Comment by cprussin** [comment]
@balazsorban44 sorry to bug you but it looks like you're a maintainer who's active in jest issues... if you can just advise me on whether the solution I'm proposing above would be acceptable to the next.js team, I'd be more than happy to put in a PR to implement the fix!

Sorry again for the tag, I know you all get a flood of messages but this one is pretty serious to me and I'm happy to fix it so I don't want the ticket to get buried in the pile.  Thank you!

**Comment by balazsorban44** [comment]
THANKS, we are open for a PR!

**Comment by cprussin** [comment]
@balazsorban44 thanks for the follow up, fixed in #47620
